support for Custom Partitioning and RAID

t0mk opened the pull request:
This PR adds a "storage" field to the device resource, where uses can specify custom partitioning and RAID setup. Packet has customer demand for this. The format of the specification is documented at https://help.packet.net/technical/storage/custom-partitioning-raid.

The storage specification is basically a JSON document specifying partitions, mountpoints and RAID setup; not unlike CoreOS ignition (https://coreos.com/ignition/docs/latest/configuration-v2_1.html), but not completely same IIUC. In Packet, also non-CoreOS devices can be partitioned.

The challenge here (from TF state point of view) was to make sure that only changes in the map structure will generate diff in TF state. Simply changing the order of how key-values are declared in the JSON document should not be detected as a change.

I normalize the JSON (json.Marshal(json.Unmarshal(jsonDoc)), so that the same maps are hopefully represented by the same JSON document. It's an approach I noticed in the aws provider. I do it through the StateFunc of the storage field.

There are no acceptance tests for this, because the `storage` can only be specified for device deployed on reserved hardware.

Fixes #34

t0mk commented:
@radeksimko Please take a look when you have time.

catsby commented on packet/provider.go:
```diff
@@ -31,6 +34,33 @@ func Provider() terraform.ResourceProvider {
 	}
 }
 
+func normalizeJSON(jsonValue interface{}) (string, error) {
```
Terraform offers a `structure` package with many methods that look like this, we encourage (but do not require) you use that

- https://github.com/hashicorp/terraform/tree/master/helper/structure

t0mk replied:
Great! Of course prefer to use inbuilts rather than new code. This taught me not to look to the aws provider for inspiration :D.

catsby commented on packet/resource_packet_device.go:
```diff
@@ -172,6 +173,23 @@ func resourcePacketDevice() *schema.Resource {
+					s, _ := normalizeJSON(v)
+					return s
+				},
+				ValidateFunc: func(v interface{}, k string) (ws []string, es []error) {
```
Terraform includes a package `validation` with a `ValidateJsonString` method you could use here

- https://github.com/hashicorp/terraform/blob/a41d5d634d282654a1c859ad4507463c25e82a00/helper/validation/validation.go#L163-L168

t0mk replied:
dtto

catsby commented on packet/resource_packet_device.go:
```diff
@@ -276,6 +302,7 @@ func resourcePacketDeviceRead(d *schema.ResourceData, meta interface{}) error {
 	d.Set("ipxe_script_url", device.IPXEScriptURL)
 	d.Set("always_pxe", device.AlwaysPXE)
 	d.Set("root_password", device.RootPassword)
+	d.Set("storage", device.Storage)
```
For any non-scalar value (e.g. `[]string` or `map[string]interface{}`), you should check the error returned from `d.Set`:

```
if err := d.Set("storage", device.Storage); err != nil {
  log.Printf("[ERR] Error setting storage for (%s): %s", d.Id(), err)
}
```

t0mk replied:
OK, will do.

t0mk commented:
@catsby could you please ACK my fixes to your review?

my commit that addresses your review is:
https://github.com/terraform-providers/terraform-provider-packet/pull/35/commits/595449fc027255f81662318298e62731463d960a
